makefiles/tools/esptool.inc.mk

```diff
@@ -10,10 +10,11 @@ BOOTLOADER_BIN = bootloader$(BOOTLOADER_COLOR)$(BOOTLOADER_INFO).bin
 ESPTOOL ?= $(RIOTTOOLS)/esptool/esptool.py
 
 # The ELFFILE is the base one used for flashing
+ELFFILE ?= $(BINDIR)/$(APPLICATION).elf
```

Comment:
Why does this need defining?

Reply:
I mean shouldn't it be known?

Reply:
Hi, it is not know at this stage.
While not having `ELFFILE` set at this point is fine if you use $(ELFFILE) in the list of commands for a rule, for example the code I'm removing had a `ls -l $(FLASHFILE).bin` command which is fine because at the time those commands are executed the `ELFFILE` variable is already set, it didn't work here because now I'm setting new rules (`$(FLASHFILE).bin: $(FLASHFILE)`) that use `ELFFILE` or `FLASHFILE` in the rule target name and dependencies, so `FLASHFILE` (and `ELFFILE`) need to be evaluated at the time this file is imported. The issue with not listing the `$(FLASHFILE)` as a dependency in the rules is that  `make -j9 flash` doesn't work as expected since the rule is executed too early.

This makefile file is imported when `PROGRAMMER` is set to `esptool`, from here: https://github.com/RIOT-OS/RIOT/blob/fed2c8e87ec0269e53388ad60c030b4f0495a266/Makefile.include#L420
but the `ELFFILE` is set in the Mafile.include a little later, from this line:
https://github.com/RIOT-OS/RIOT/blob/fed2c8e87ec0269e53388ad60c030b4f0495a266/Makefile.include#L559
so at this point `ELFFILE` is not set.

I don't think this is a problem since the `esptool` programmer is really only useful for the ESP CPUs, so having the default assumption that we are flashing an ELFFILE with this name is fine. This is also a `?=` assignment so the board or app can override `ELFFILE` if they want to.

Reply:
Ok that makes sense, in that case I think that we can add it there, although it seems moving the PROGRAMMER include later is the best fix but is a huge change that would require more testing. @iosabi can you add a comment summarizing what you just described just before `ELFFILE ?= $(BINDIR)/$(APPLICATION).elf`?

Reply:
I'll add a comment to that line.